### PR TITLE
add massive migrations & L2 maxNumBatch

### PR DESCRIPTION
**Pull request by krlosMata**
- This PR adds the following features into the protocol:
  - allow massive migrations by adding L1 transaction data to data availability
  - add timeout for L2 transactions measured in batch numbers

## Massive migrations
- `hermez zkRollup protocol`
  - update `data-availability`
- `circuits`
  - compute `isAmountNullified` in `balance-updater`
  - compute `effectiveAmountF = amountF * (1 - isAmountNullified)` in `rollup-main`

## L2 timeout
- `hermez zkRollup protocol`
  - add new global variable `CURRENT_BATCH_NUMBER`
  - add new field `maxNumBatch` to be signed in L2 transactions 
- `circuits`
  - add `CURRENT_BATCH_NUMBER` in `rollup-main`
  - add `maxNumBatch` for each transaction
  - compare `maxNumBatch` with `CURRENT_BATCH_NUMBER` in `decode-tx`

Please, note that not all the content needs to be reviewed by all reviewers:
@ed255 & @invocamanman & @arnaucube to check protocol changes
@arnaucube to check style and format
@jbaylina to check circuits updates

### CODE REVIEW

**Review by jesusholoo — APPROVED** [review]
nice :D

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 1]
```diff
@@ -195,47 +195,50 @@ Take the transaction data, decodes it and build data structures to be used in fu
...
-| fromBjjCompressed[256] | boolean array |              babyjubjub compressed sender              |
-|      loadAmountF       |    uint16     |   amount to deposit from L1 to L2 encoded as float16   |
-|     globalChainID      |    uint16     |                global chain identifier                 |
-|        onChain         |     bool      |        determines if the transacion is L1 or L2        |
-|       newAccount       |     bool      |    determines if transaction creates a new account     |
-|       auxFromIdx       |    uint48     |           auxiliary index to create accounts           |
-|         inIdx          |    uint48     |                old last index assigned                 |
+|         Input          |     type      |                            Description                             |
+|:----------------------:|:-------------:|:------------------------------------------------------------------:|
+|    previousOnChain     |     bool      |              dtermines if previous transaction is L1               |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
|    previousOnChain     |     bool      |              determines if previous transaction is L1               |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 2]
```diff
@@ -195,47 +195,50 @@ Take the transaction data, decodes it and build data structures to be used in fu
...
+|        toBjjAy         |     field     |                  bayjubjub y coordinate receiver                   |
+|  rqTxCompressedDataV2  |    uint193    |       requested encode transaction fields together version 2       |
+|      rqToEthAddr       |    uint160    |                requested ethereum address receiver                 |
+|       rqToBjjAy        |     field     |                 requested babyjubjub y coordinate                  |
+|      fromEthAddr       |    uint160    |                      ethereum address sender                       |
+| fromBjjCompressed[256] | boolean array |                    babyjubjub compressed sender                    |
+|      loadAmountF       |    uint16     |         amount to deposit from L1 to L2 encoded as float16         |
+|     globalChainID      |    uint16     |                      global chain identifier                       |
+|    currentNumBatch     |    uint32     |                        current batch number                        |
+|        onChain         |     bool      |              determines if the transacion is L1 or L2              |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
|        onChain         |     bool      |              determines if the transaction is L1 or L2              |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 3]
```diff
@@ -432,7 +436,8 @@ This circuit checks if there is enough balance in the sender account to do the t
...
 
-It should be noted that in L1 tx, no errors are allowed but the circuit needs to process them. Hence, in case it is not enough balance on the sender account, it will process the transaction as a 0 amount transfer. In case of an L2 tx, the protocol does not allow to do a transaction if there is not enough balance on the sender account.
+It should be noted that in L1 tx, no errors are allowed but the circuit needs to process them. Hence, in case it is not enough balance on the sender account, it will process the transaction as a 0 amount transfer. Hence, signal `isAmountNullified` will notify if a L1 transaction has been nullified if it is invalid. This `isAmountNullified` will be used to compute data-availability where the amount used would not be inserted in [`L1L2TxsData`](developers/protocol/hermez-protocol/protocol?id=l1-l2-transactions) since L1Tx is not valid or triggers underflow.
+In case of an L2 tx, the protocol does not allow to do a transaction if there is not enough balance on the sender account.
 
 - The following assumptions has been taken:
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
- The following assumptions have been taken:
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 4]
```diff
@@ -457,7 +462,7 @@ It should be noted that in L1 tx, no errors are allowed but the circuit needs to
 |        amount        | uint192 |              amount to transfer from L2 to L2               |
 |      loadAmount      | uint192 |               amount to deposit from L1 to L2               |
 |     feeSelector      |  uint8  |                      user selector fee                      |
-|       onChain        |  bool   |          determines if the transaction is L1 or L2           |
+|       onChain        |  bool   |          determines if the transacion is L1 or L2           |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]: (this one was already correct previous to this PR)
```suggestion
|       onChain        |  bool   |          determines if the transaction is L1 or L2           |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 5]
```diff
@@ -721,12 +731,13 @@ Main circuit could be splitted in the following phases:
...
-|        auxToIdx[nTx]         |   uint48 array    |        auxiliary index when signed index receiver is set to null        |
+|         fromIdx[nTx]         |   uint40 array    |                              index sender                               |
+|       auxFromIdx[nTx]        |   uint40 array    |                   auxiliary index to create accounts                    |
+|          toIdx[nTx]          |   uint40 array    |                             index receiver                              |
+|        auxToIdx[nTx]         |   uint40 array    |        auxiliary index when signed index receiver is set to null        |
 |         toBjjAy[nTx]         |    field array    |                     bayjubjub y coordinate receiver                     |
 |        toEthAddr[nTx]        |   uint160 array   |                        ethereum address receiver                        |
+|       maxNumBatch[nTx]       |   uint32 array    |   maximum allowed batch number when the transaction can be processed    |
 |         onChain[nTx]         |    bool array     |                determines if the transacion is L1 or L2                 |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
|         onChain[nTx]         |    bool array     |                determines if the transaction is L1 or L2                 |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 6]
```diff
@@ -721,12 +731,13 @@ Main circuit could be splitted in the following phases:
...
-|         fromIdx[nTx]         |   uint48 array    |                              index sender                               |
-|       auxFromIdx[nTx]        |   uint48 array    |                   auxiliary index to create accounts                    |
-|          toIdx[nTx]          |   uint48 array    |                             index receiver                              |
-|        auxToIdx[nTx]         |   uint48 array    |        auxiliary index when signed index receiver is set to null        |
+|         fromIdx[nTx]         |   uint40 array    |                              index sender                               |
+|       auxFromIdx[nTx]        |   uint40 array    |                   auxiliary index to create accounts                    |
+|          toIdx[nTx]          |   uint40 array    |                             index receiver                              |
+|        auxToIdx[nTx]         |   uint40 array    |        auxiliary index when signed index receiver is set to null        |
 |         toBjjAy[nTx]         |    field array    |                     bayjubjub y coordinate receiver                     |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
|         toBjjAy[nTx]         |    field array    |                     babyjubjub y coordinate receiver                     |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 7]
```diff
@@ -195,47 +195,50 @@ Take the transaction data, decodes it and build data structures to be used in fu
...
-|       newAccount       |     bool      |    determines if transaction creates a new account     |
-|       auxFromIdx       |    uint48     |           auxiliary index to create accounts           |
-|         inIdx          |    uint48     |                old last index assigned                 |
+|         Input          |     type      |                            Description                             |
+|:----------------------:|:-------------:|:------------------------------------------------------------------:|
+|    previousOnChain     |     bool      |              dtermines if previous transaction is L1               |
+|    txCompressedData    |    uint241    |                 encode transaction fields together                 |
+|      maxNumBatch       |    uint32     | maximum allowed batch number when the transaction can be processed |
+|       toEthAddr        |    uint160    |                     ethereum address receiver                      |
+|        toBjjAy         |     field     |                  bayjubjub y coordinate receiver                   |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
|        toBjjAy         |     field     |                  babyjubjub y coordinate receiver                   |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 8]
```diff
@@ -700,19 +707,22 @@ Main circuit could be splitted in the following phases:
...
+|       currentNumBatch        |      uint32       |                     current batch number processed                      |
+|      feeIdxs[maxFeeTx]       |   uint40 array    |                   merkle tree indexes to receive fees                   |
 |   feePlanTokens[maxFeeTx]    |   uint32 array    |                 tokens identifiers of fees accumulated                  |
 |       imOnChain[nTx-1]       |   boolean array   |      intermediary signals: decode transaction output onChain flag       |
-|       imOutIdx[nTx-1]        |   uint48 array    |      intermediary signals: decode transaction final index assigned      |
+|       imOutIdx[nTx-1]        |   uint40 array    |      intermediary signals: decode transaction final index assigned      |
 |      imStateRoot[nTx-1]      |    field array    |           intermediary signals: transaction final state root            |
 |      imExitRoot[nTx-1]       |    field array    |            intermediary signals: transaction final exit root            |
 | imAccFeeOut[nTx-1][maxFeeTx] |   uint192 array   |         intermediary signals: transaction final accumlate fees          |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]:
```suggestion
| imAccFeeOut[nTx-1][maxFeeTx] |   uint192 array   |         intermediary signals: transaction final accumulate fees          |
```

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md [hunk 9]
```diff
@@ -756,15 +767,15 @@ Main circuit could be splitted in the following phases:
...
 |        oldValue2[nTx]        |    field array    |                      old value of the sender leaf                       |
 |      tokenID3[maxFeeTx]      |   uint32 array    |                        tokenID of leafs feeIdxs                         |
 |       nonce3[maxFeeTx]       |   uint40 array    |                         nonce of leafs feeIdxs                          |
 |       sign3[maxFeeTx]        |    bool array     |                          sign of leafs feeIdxs                          |
 |      balance3[maxFeeTx]      |   uint192 array   |                        balance of leafs feeIdxs                         |
 |        ay3[maxFeeTx]         |    field array    |                           ay of leafs feeIdxs                           |
 |      ethAddr3[maxFeeTx]      |   uint160 array   |                        ethAddr of leafs feeIdxs                         |
-|    siblings3[nLevels + 1]    | field array       |                   siblings merkle proof of leafs Idxs                   |
+|    siblings3[nLevels + 1]    | field array array |                   siblings merkle proof of leafs Idxs                   |
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/circuits/circuits.md]: Missing `[maxFeeIdxs]`. Apart, at the 2nd column appears the `field array array`, while at the same column for the row of `siblings1` and `siblings2` is also an *array of arrays of field elements* but only appears as `field array` (lines 758 and 768) (would affect also parameter descriptions such as the `FromBJJCompressed`)
```suggestion
|    siblings3[maxFeeIdxs][nLevels + 1]    | field array array |                   siblings merkle proof of leafs Idxs                   |
```

File: docs/developers/protocol/hermez-protocol/protocol.md [hunk 10]
```diff
@@ -695,18 +702,18 @@ $len(l1\_txs) \leq MAX\_L1\_TXS < MAX\_TXS$
 $len(l1\_user\_txs) \leq MAX\_L1\_USER\_TXS < MAX\_L1\_TXS$
 
 ### L1 user transactions
-All transaction data triggered by a smart contract function can be directly retrieved since it will be stored on the blockchain, but it's harder when this data happens in internal transactions, not all nodes support that functionality. That's why all the L1 user transactions emit an L1UserTx event to facilitate the data retrieval.
+All transaction data triggered by a smart contract function can be directly retrieved since it will be stored on the blockchain, but it's harder when this data happens in internal transactions, not all nodes support that functionality. That's why all the L1 user transactions emit an `L1UserTx` event to facilitate the data retrieval.
 
-When a user calls a function that adds an L1UserTx, the folowing happens:
+When a user calls a function that adds an `L1UserTx`, the folowing happens:
```
Comment (arnaucube) [inline docs/developers/protocol/hermez-protocol/protocol.md]:
```suggestion
When a user calls a function that adds an `L1UserTx`, the following happens:
```

**Review by arnaucube — CHANGES_REQUESTED** [review]

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md (hunk 4 above)
Comment (krlosMata) [reply docs/developers/protocol/hermez-protocol/circuits/circuits.md]: Update has been made with an old version table.
Fixed.

File: docs/developers/protocol/hermez-protocol/circuits/circuits.md (hunk 9 above)
Comment (krlosMata) [reply docs/developers/protocol/hermez-protocol/circuits/circuits.md]: Added `array array` in all inputs consisting in *array of arrays of field elements*

File: docs/developers/protocol/hermez-protocol/protocol.md [hunk 11]
```diff
@@ -95,6 +95,7 @@ dataBuffer: [48 bits] fromIdx
 - `MAX_TX`: absolute maximum L1 or L2 transactions allowed to process in one batch
 - `MAX_L1_TXS`: absolute maximum of L1 transactions allowed to process in one batch
 - `CHAIN_ID`: identifier zkRollup chain
+- `CURRENT_BATCH_NUMBER`: current batch number processed
```
Comment (ed255) [inline docs/developers/protocol/hermez-protocol/protocol.md]: Why is this variable in uppercase?  I understand that:
1. It's not a constant
2. It's not a parameter of the circuit (but an input).

Also, why is it under Global Settings?

**Review by ed255 — CHANGES_REQUESTED** [review]

File: docs/developers/protocol/hermez-protocol/protocol.md (hunk 11 above)
Comment (krlosMata) [reply docs/developers/protocol/hermez-protocol/protocol.md]: it is an input forced by the smart contract, so I think it fits on global settings but in `contract` section. (same applies for `chain_id` which should be also in lowercase).

**Review by jbaylina — APPROVED** [review]

**Review by ed255 — APPROVED** [review]

**Review by arnaucube — APPROVED** [review]

**Pull request merged**